modules/navier_stokes/src/fvkernels/PCNSFVHLLC.C

```diff
@@ -0,0 +1,139 @@
+    _rho_et_neighbor(getNeighborADMaterialProperty<Real>(NS::total_energy_density)),
+    _vel_elem(getADMaterialProperty<RealVectorValue>(NS::velocity)),
+    _vel_neighbor(getNeighborADMaterialProperty<RealVectorValue>(NS::velocity)),
+    _speed_elem(getADMaterialProperty<Real>(NS::speed)),
+    _speed_neighbor(getNeighborADMaterialProperty<Real>(NS::speed)),
+    _rho_elem(getADMaterialProperty<Real>(NS::density)),
+    _rho_neighbor(getNeighborADMaterialProperty<Real>(NS::density)),
+    _pressure_elem(getADMaterialProperty<Real>(NS::pressure)),
+    _pressure_neighbor(getNeighborADMaterialProperty<Real>(NS::pressure)),
+    _eps_elem(getMaterialProperty<Real>(NS::porosity)),
+    _eps_neighbor(getNeighborMaterialProperty<Real>(NS::porosity))
+{
+}
+
+std::vector<ADReal>
+PCNSFVHLLC::waveSpeed(const ADReal & rho_elem,
+                      const ADRealVectorValue & vel_elem,
+                      const ADReal & e_elem,
+                      const Real eps_elem,
+                      const ADReal & rho_neighbor,
+                      const ADRealVectorValue & vel_neighbor,
+                      const ADReal & e_neighbor,
+                      const Real eps_neighbor,
+                      const SinglePhaseFluidProperties & fluid,
+                      const ADRealVectorValue & normal)
+{
+  const auto & rho1 = rho_elem;
+  const auto u1 = vel_elem.norm();
+  const auto q1 = normal * vel_elem;
+  const auto v1 = 1.0 / rho1;
+  const auto & e1 = e_elem;
+  const auto et1 = e1 + 0.5 * u1 * u1;
+  const auto p1 = fluid.p_from_v_e(v1, e1);
+  const auto ht1 = et1 + p1 / rho1;
+  const auto c1 = fluid.c_from_v_e(v1, e1);
+  const auto eps1 = eps_elem;
+
+  const auto & rho2 = rho_neighbor;
+  const auto u2 = vel_neighbor.norm();
+  const auto q2 = normal * vel_neighbor;
+  const auto v2 = 1.0 / rho2;
+  const auto & e2 = e_neighbor;
+  const auto et2 = e2 + 0.5 * u2 * u2;
+  const auto p2 = fluid.p_from_v_e(v2, e2);
+  const auto ht2 = et2 + p2 / rho2;
+  const auto c2 = fluid.c_from_v_e(v2, e2);
+  const auto eps2 = eps_neighbor;
+
+  // compute Roe-averaged variables
+  const auto sqrt_rho1 = std::sqrt(rho1);
+  const auto sqrt_rho2 = std::sqrt(rho2);
+  const auto u_roe = (sqrt_rho1 * u1 + sqrt_rho2 * u2) / (sqrt_rho1 + sqrt_rho2);
+  const auto q_roe = (sqrt_rho1 * q1 + sqrt_rho2 * q2) / (sqrt_rho1 + sqrt_rho2);
+  const auto e_roe = (sqrt_rho1 * e1 + sqrt_rho2 * e2) / (sqrt_rho1 + sqrt_rho2);
+  const auto rho_roe = std::sqrt(rho1 * rho2);
+  const auto v_roe = 1.0 / rho_roe;
+  const auto c_roe = fluid.c_from_v_e(v_roe, e_roe);
+
+  // compute wave speeds
+  // I may want to change the estimate of these wave speeds!
+  const auto SL = std::min(q1 - c1, q_roe - c_roe);
+  const auto SR = std::max(q2 + c2, q_roe + c_roe);
+  const auto SM =
+      (eps2 * rho2 * q2 * (SR - q2) - eps1 * rho1 * q1 * (SL - q1) + eps1 * p1 - eps2 * p2) /
+      (eps2 * rho2 * (SR - q2) - eps1 * rho1 * (SL - q1));
+
+  // store these results in _wave_speed
```

Comment:
you're not caching them for PCNSFV?

Reply:
No. I suppose I should make them consistent, huh. Ugh I suppose I'll profile and see if it matters

Reply:
Ok I profiled a PCNSFV simulation and with caching `PCNSFVHLLC::waveSpeed` comprises 6.44% of the simulation time. Without caching it comprises 6.68%. Comparatively speaking, `MooseVariableDataFV::computeAD` takes ~55% of the simulation time.

Given these results, I don't really care what we do. Maybe I'll do a heap profile and see whether the cache is noticeable

Reply:
Oops I added the cache but didn't actually use it :rofl: Stand-by for actual caching CPU and heap profile results...

Reply:
`PCNSFVHLLC::waveSpeed`:

CPU:
- caching: 3.50%
- not caching: 6.68%

HEAP:
- cache takes anywhere between 10% and 25% of total memory depending on the snapshot

My thought would be that the increase in memory use is not worth the 3% drop in CPU. @GiudGiud what do you think?

Reply:
No more caching